File: src/go/fab/internal/hooks/hooks.go
package hooks

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

// Run executes a shell command string in the given working directory.
// The command is run via "sh -c" so it supports pipes, redirects, etc.
// If command is empty, returns nil (no-op).
// The working directory is the repo root (parent of fabRoot).
func Run(fabRoot, command string) error {
	if command == "" {
		return nil
	}

	repoRoot := filepath.Dir(fabRoot)

	// Support script paths: if the command is a path to a file, resolve it
	// relative to the repo root.
	cmd := exec.Command("sh", "-c", command)
	cmd.Dir = repoRoot
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("stage hook failed: %s: %w", command, err)
	}
	return nil
}

File: src/fab-rust/src/hooks.rs
use anyhow::{bail, Result};
use std::path::Path;
use std::process::Command;

/// Executes a shell command string in the repo root directory.
/// The command is run via "sh -c" so it supports pipes, redirects, etc.
/// If command is empty, returns Ok (no-op).
pub fn run(fab_root: &str, command: &str) -> Result<()> {
    if command.is_empty() {
        return Ok(());
    }

    let repo_root = Path::new(fab_root)
        .parent()
        .unwrap_or(Path::new("/"));

    let status = Command::new("sh")
        .arg("-c")
        .arg(command)
        .current_dir(repo_root)
        .status()?;

    if !status.success() {
        bail!("stage hook failed: {}", command);
    }
    Ok(())
}
